File: src/main/java/gargoyle/netsquares/model/Game.java
package gargoyle.netsquares.model;

import gargoyle.netsquares.res.AudioClip;
import gargoyle.netsquares.res.Resources;
import gargoyle.netsquares.store.Options;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

public final class Game implements Serializable, Cloneable {
    private static final String RES_BG = "sound/bg/au/Frunze-Evenings.au";
    private static final String RES_TAP = "sound/fx/tap.wav";
    private final Board board;
    private final AudioClip mus;
    private final Options options;
    private final AudioClip tap;
    private int current;
    private Map<Player, Directions> playerDirectionsMap;
    private final List<Player> players;
    private GameState state;

    public Game(Options options) {
        this.options = options;
        state = GameState.MENU;
        board = new Board();
        players = new ArrayList<>();
        playerDirectionsMap = new HashMap<>();
        tap = Resources.loadAudio(RES_TAP, Locale.getDefault());
        mus = Resources.loadAudio(RES_BG, Locale.getDefault());
    }

    private Game(Options options, Board board, Player... players) {
        this(options, board, Arrays.asList(players));
    }

    private Game(Options options, Board board, @NotNull List<Player> players) {
        this.options = options;
        this.board = board;
        this.state = GameState.GAME;
        this.players = new ArrayList<>();
        initPlayers(players);
        tap = Resources.loadAudio(RES_TAP, Locale.getDefault());
        mus = Resources.loadAudio(RES_BG, Locale.getDefault());
    }

    public static @NotNull Game createGame(Options options, Board board, Player... players) {
        return new Game(options, board, players).beginGame();
    }

    private void beginGame(Player... players) {
        initPlayers(Arrays.asList(players));
        board.reInit();
        state = GameState.GAME;
        if (options.isSound() && options.isMusic()) {
            musicStart();
        }
    }

    @NotNull
    private Game beginGame() {
        resetPlayers();
        board.reInit();
        state = GameState.GAME;
        return this;
    }

    public void beginGame(int boardSize, int boardWeight, Player... players) {
        board.init(boardSize, boardWeight);
        beginGame(players);
    }

    private boolean canMove() {
        return canMove(getCurrentPlayer());
    }

    @SuppressWarnings("LawOfDemeter")
    private boolean canMove(Player player) {
        if (board.canOpen()) {
            return true;
        }
        Directions directions = getAllowed(player);
        for (Direction direction : directions.getDirections()) {
            for (int i = 1; board.canMoveBy(direction, i); i++) {
                if (board.canOpenBy(direction, i)) {
                    return true;
                }
            }
        }
        return false;
    }

    public @Nullable Player computeWinner() {
        int maxScore = Integer.MIN_VALUE;
        Player winner = null;
        for (Player player : players) {
            int score = player.getScore();
            if (maxScore < score) {
                maxScore = score;
                winner = player;
            }
        }
        return winner;
    }

    public void endGame() {
        if (GameState.GAME.isInState(state)) {
            state = GameState.OPEN;
            board.openBoard();
        }
        musicStop();
    }

    public Directions getAllowed(@NotNull Player player) {
        return playerDirectionsMap.get(player);
    }

    public Board getBoard() {
        return board;
    }

    public Player getCurrentPlayer() {
        return players.isEmpty() ? null : players.get(current);
    }

    public List<Player> getPlayers() {
        //noinspection AssignmentOrReturnOfFieldWithMutableType
        return players;
    }

    public GameState getState() {
        return state;
    }

    public void setState(GameState state) {
        this.state = state;
    }

    @Override
    public int hashCode() {
        return Objects.hash(board, players, current, state);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Game game = (Game) o;
        return current == game.current &&
                Objects.equals(board, game.board) &&
                Objects.equals(players, game.players) &&
                state == game.state;
    }

    @Override
    public @NotNull Game clone() throws CloneNotSupportedException {
        return (Game) super.clone();
    }

    @Override
    public String toString() {
        return String.format("Game{state=%s, current=%d, playerDirectionsMap=%s, players=%s, board=%s}", state, current, playerDirectionsMap, players, board);
    }

    private void initPlayers(@NotNull List<Player> players) {
        this.players.clear();
        this.players.addAll(players);
        Map<Player, Directions> playerDirectionsMap = new HashMap<>();
        for (int i = 0; i < players.size(); i++) {
            Player player = players.get(i);
            Directions directions = i % 2 == 0 ? Directions.HORIZONTALS : Directions.VERTICALS;
            playerDirectionsMap.put(player, directions);
        }
        this.playerDirectionsMap = Collections.unmodifiableMap(playerDirectionsMap);
    }

    public boolean isAllowed(Player player, Direction direction) {
        return direction.isAllowed(getAllowed(player));
    }

    public boolean isState(GameState state) {
        return this.state == state;
    }

    public boolean isState(GameState... states) {
        return this.state.isInState(states);
    }

    @SuppressWarnings("LawOfDemeter")
    private void makeMove() {
        Player currentPlayer = getCurrentPlayer();
        if (currentPlayer == null) {
            return;
        }
        Move move = currentPlayer.move(board, getAllowed(currentPlayer));
        if (move == null) {
            return;
        }
        if (options.isSound()) {
            tap.play();
        }
        if (move == Move.OPEN) {
            currentPlayer.addScore(board.openAtCaret());
            nextPlayer();
            return;
        }
        if (getAllowed(currentPlayer).isAllowed(move.getDirection()) && board.canMoveBy(move)) {
            board.moveBy(move);
        }
    }

    public boolean move() {
        if (GameState.GAME.isInState(state) && canMove()) {
            makeMove();
            return true;
        }
        return false;
    }

    public void musicStart() {
        mus.loop();
    }

    public void musicStop() {
        mus.stop();
    }

    private void nextPlayer() {
        current = (current + 1) % players.size();
    }

    private void resetPlayers() {
        for (Player player : players) {
            player.reset();
        }
    }
}

File: src/main/kotlin/gargoyle/netsquares/model/Game.kt
package gargoyle.netsquares.model

import gargoyle.netsquares.res.AudioClip
import gargoyle.netsquares.res.NSContext
import gargoyle.netsquares.store.Options

class Game private constructor(
    context: NSContext,
    private val options: Options,
    val board: Board,
    players: List<Player>,
    var state: GameState
) {
    private val mus: AudioClip?
    private val players: MutableList<Player>
    private val tap: AudioClip?
    private var current = 0
    private var playerDirectionsMap: Map<Player?, Directions>? = null

    constructor(context: NSContext, options: Options) : this(context, options, Board(), GameState.MENU)
    private constructor(
        context: NSContext,
        options: Options,
        board: Board,
        state: GameState,
        vararg players: Player
    ) : this(context, options, board, listOf<Player>(*players), state)

    init {
        this.players = mutableListOf()
        tap = context.loadAudio(RES_TAP, "wav")
        mus = context.loadAudio(RES_BG, "au")
        initPlayers(players)
    }

    private fun initPlayers(players: List<Player>) {
        this.players.clear()
        this.players.addAll(players)
        val playerDirectionsMap: MutableMap<Player?, Directions> = mutableMapOf()
        for (i in players.indices) {
            val player = players[i]
            val directions = if (i % 2 == 0) Directions.HORIZONTALS else Directions.VERTICALS
            playerDirectionsMap[player] = directions
        }
        this.playerDirectionsMap = playerDirectionsMap
    }

    private fun beginGame(): Game {
        resetPlayers()
        board.reInit()
        state = GameState.GAME
        return this
    }

    private fun beginGame(vararg players: Player) {
        initPlayers(listOf(*players))
        board.reInit()
        state = GameState.GAME
        if (options.sound.get() && options.music.get()) {
            musicStart()
        }
    }

    fun beginGame(boardSize: Int, boardWeight: Int, vararg players: Player) {
        board.init(boardSize, boardWeight)
        beginGame(*players)
    }

    fun musicStart() {
        mus!!.loop()
    }

    fun computeWinner(): Player? {
        var maxScore = Int.MIN_VALUE
        var winner: Player? = null
        for (player in players) {
            val score = player.score
            if (maxScore < score) {
                maxScore = score
                winner = player
            }
        }
        return winner
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val game = other as Game
        return current == game.current &&
                board == game.board &&
                players == game.players && state == game.state
    }

    fun endGame() {
        if (GameState.GAME.isInState(state)) {
            state = GameState.OPEN
            board.openBoard()
        }
        musicStop()
    }

    fun getPlayers(): List<Player> {
        return players
    }

    override fun hashCode(): Int {
        return arrayOf(board, players, current, state).contentHashCode()
    }

    fun move(): Boolean {
        if (GameState.GAME.isInState(state) && canMove()) {
            makeMove()
            return true
        }
        return false
    }

    override fun toString(): String {
        return String.format(
            "Game{state=%s, current=%d, playerDirectionsMap=%s, players=%s, board=%s}",
            state,
            current,
            playerDirectionsMap,
            players,
            board
        )
    }

    fun isAllowed(player: Player, direction: Direction): Boolean {
        return direction.isAllowed(getAllowed(player))
    }

    private fun canMove(player: Player = currentPlayer!!): Boolean {
        if (board.canOpen()) {
            return true
        }
        val directions = getAllowed(player)
        for (direction in directions.getDirections()) {
            var i = 1
            while (board.canMoveBy(direction, i)) {
                if (board.canOpenBy(direction, i)) {
                    return true
                }
                i++
            }
        }
        return false
    }

    fun isState(state: GameState): Boolean {
        return this.state == state
    }

    fun isState(vararg states: GameState): Boolean {
        return state.isInState(*states)
    }

    val currentPlayer: Player?
        get() = if (players.isEmpty()) null else players[current]

    fun getAllowed(player: Player): Directions {
        return playerDirectionsMap!![player]!!
    }

    private fun makeMove() {
        val currentPlayer = currentPlayer ?: return
        val move = currentPlayer.move(board, getAllowed(currentPlayer)) ?: return
        if (options.sound.get()) {
            tap!!.play()
        }
        if (move == Move.OPEN) {
            currentPlayer.addScore(board.openAtCaret())
            nextPlayer()
            return
        }
        if (getAllowed(currentPlayer).isAllowed(move.direction) && board.canMoveBy(move)) {
            board.moveBy(move)
        }
    }

    fun musicStop() {
        mus!!.stop()
    }

    private fun resetPlayers() {
        for (player in players) {
            player.reset()
        }
    }

    private fun nextPlayer() {
        current = (current + 1) % players.size
    }

    companion object {
        private const val RES_BG = "gargoyle/netsquares/res/Frunze-Evenings"
        private const val RES_TAP = "gargoyle/netsquares/res/tap"
        fun createGame(context: NSContext, options: Options, board: Board, vararg players: Player): Game {
            return Game(context, options, board, GameState.GAME, *players).beginGame()
        }
    }
}
